File: tests/python/test-RNA-boltzmann-sampling.py
import RNApath

RNApath.addSwigInterfacePath()

import RNA
import unittest

sequence = "UGGGAAUAGUCUCUUCCGAGUCUCGCGGGCGACGGGCGAUCUUCGAAAGUGGAAUCCGUACUUAUACCGCCUGUGCGGACUACUAUCCUGACCACAUAGU"

def store_structure(s, data):
    if s:
        data.append(s)


def prepare_fc():
    md          = RNA.md()
    md.uniq_ML  = 1
    fc          = RNA.fold_compound(sequence, md)
    (ss, mfe)   = fc.mfe()

    fc.exp_params_rescale(mfe)
    fc.pf()

    return fc

class GeneralTests(unittest.TestCase):
    def test_pbacktrack5(self):
        print "test_pbacktrack    (single sub-structure, a.k.a. pbacktrack5)"
        fc = prepare_fc()

        s  = fc.pbacktrack(10)
        self.assertEqual(len(s), 10)

        s  = fc.pbacktrack(50)
        self.assertEqual(len(s), 50)

    def test_pbacktrack5_multi(self):
        print "test_pbacktrack    (multiple sub-structures, a.k.a. pbacktrack5_num)"
        fc = prepare_fc()

        ss  = fc.pbacktrack(10, 20)
        self.assertEqual(len(ss), 20)
        for s in ss:
            self.assertEqual(len(s), 10)

        ss  = fc.pbacktrack(50, 100)
        self.assertEqual(len(ss), 100)
        for s in ss:
            self.assertEqual(len(s), 50)

    def test_pbacktrack(self):
        print "test_pbacktrack    (single structure, a.k.a. pbacktrack)"
        fc = prepare_fc()

        s  = fc.pbacktrack()
        self.assertEqual(len(s), len(sequence))

    def test_pbacktrack_multi(self):
        print "test_pbacktrack    (multiple structures, a.k.a. pbacktrack_num)"
        fc = prepare_fc()

        ss  = fc.pbacktrack(0, 100)
        self.assertEqual(len(ss), 100)
        for s in ss:
            self.assertEqual(len(s), len(sequence))

    def test_pbacktrack_nr(self):
        print "test_pbacktrack_nr"
        fc = prepare_fc()

        ss  = fc.pbacktrack_nr(100)
        self.assertEqual(len(ss), 100)
        for s in ss:
            self.assertEqual(len(s), len(sequence))
        # check for uniqueness, i.e. no duplicates
        sss = list(set(ss))
        self.assertEqual(len(ss), len(sss))

    def test_pbacktrack_nr_resume(self):
        print "test_pbacktrack_nr_resume"
        num_samples = 500
        iterations  = 5
        fc      = prepare_fc()
        d       = None
        s_list  = list()
        for i in range(0, iterations):
            d, ss   = fc.pbacktrack_nr_resume(num_samples, d)
            s_list  = s_list + list(ss)

        self.assertEqual(len(s_list), iterations * num_samples)

        for s in s_list:
            self.assertEqual(len(s), len(sequence))

        # check for uniqueness, i.e. no duplicates
        sss = list(set(s_list))
        self.assertEqual(len(s_list), len(sss))

    def test_pbacktrack5_cb(self):
        print "test_pbacktrack_cb (multiple sub-structures, a.k.a. pbacktrack5_cb)"
        fc = prepare_fc()
        ss = list()
        i = fc.pbacktrack_cb(10, 20, store_structure, ss)
        self.assertEqual(i, 20)
        self.assertEqual(len(ss), 20)
        for s in ss:
            self.assertEqual(len(s), 10)

        ss = list()
        i = fc.pbacktrack_cb(50, 100, store_structure, ss)
        self.assertEqual(i, 100)
        self.assertEqual(len(ss), 100)
        for s in ss:
            self.assertEqual(len(s), 50)

    def test_pbacktrack_cb(self):
        print "test_pbacktrack_cb (multiple structures, a.k.a. pbacktrack_cb)"
        fc = prepare_fc()
        ss = list()
        i = fc.pbacktrack_cb(0, 100, store_structure, ss)
        self.assertEqual(i, 100)
        self.assertEqual(len(ss), 100)
        for s in ss:
            self.assertEqual(len(s), len(sequence))

    def test_pbacktrack_nr_cb(self):
        print "test_pbacktrack_nr_cb"
        fc = prepare_fc()
        ss = list()
        i = fc.pbacktrack_nr_cb(100, store_structure, ss)
        self.assertEqual(i, 100)
        self.assertEqual(len(ss), 100)
        for s in ss:
            self.assertEqual(len(s), len(sequence))

        # check for uniqueness, i.e. no duplicates
        sss = list(set(ss))
        self.assertEqual(len(ss), len(sss))

    def test_pbacktrack_nr_resume_cb(self):
        print "test_pbacktrack_nr_resume_cb"
        num_samples = 500
        iterations  = 5
        fc = prepare_fc()
        ss = list()
        d  = None
        for i in range(0, iterations):
            d, i = fc.pbacktrack_nr_resume_cb(num_samples, store_structure, ss, d)
            self.assertEqual(i, num_samples)

        self.assertEqual(len(ss), iterations * num_samples)
        for s in ss:
            self.assertEqual(len(s), len(sequence))

        # check for uniqueness, i.e. no duplicates
        sss = list(set(ss))
        self.assertEqual(len(ss), len(sss))


if __name__ == '__main__':
    unittest.main();

File: tests/perl5/test-RNA-boltzmann-sampling.pl
use strict;
use warnings;
use Test::More tests => 32;

use RNA;

my $sequence = "UGGGAAUAGUCUCUUCCGAGUCUCGCGGGCGACGGGCGAUCUUCGAAAGUGGAAUCCGUACUUAUACCGCCUGUGCGGACUACUAUCCUGACCACAUAGU";

sub store_structure {
  my $s     = shift;
  my $data  = shift;
  push @{$data}, $s if defined($s);
}


sub prepare_fc {
  my $md = new RNA::md();
  $md->{uniq_ML} = 1;
  my $fc = new RNA::fold_compound($sequence, $md);

  my ($ss, $mfe) = $fc->mfe();

  $fc->exp_params_rescale($mfe);
  $fc->pf();

  return $fc;
}


sub uniq {
  my %seen;
  grep !$seen{$_}++, @_;
}


####################
# Begin actual tests
####################
my $fc          = prepare_fc();
my $failure     = 0;
my $num_samples = 500;
my $iterations  = 15;
my $s;
my $i;
my @ss;
my @sss;
my $d;

print "test_pbacktrack5    (single sub-structure, a.k.a. pbacktrack5)\n";
$s = $fc->pbacktrack5(10);
ok(length($s) == 10);

$s = $fc->pbacktrack5(50);
ok(length($s) == 50);


print "test_pbacktrack5    (multiple sub-structures, a.k.a. pbacktrack5_num)\n";
$s = $fc->pbacktrack5(20, 10);

ok(scalar(@{$s}) == 20);

$failure = 0;
foreach my $s (@{$s}) {
  if (length($s) != 10) {
    $failure = 1;
    last;
  }
}
ok($failure == 0);

$s = $fc->pbacktrack5(100, 50);
ok(scalar(@{$s}) == 100);

$failure = 0;
foreach my $s (@{$s}) {
  if (length($s) != 50) {
    $failure = 1;
    last;
  }
}
ok($failure == 0);


print "test_pbacktrack     (single structure, a.k.a. pbacktrack)\n";
$s = $fc->pbacktrack();
ok(length($s) == length($sequence));


print "test_pbacktrack     (multiple structures, a.k.a. pbacktrack_num)\n";
$s = $fc->pbacktrack(100);
ok(scalar(@{$s}) == 100);

$failure = 0;
foreach my $s (@{$s}) {
  if (length($s) != length($sequence)) {
    $failure = 1;
    last;
  }
}
ok($failure == 0);


print "test_pbacktrack_nr\n";
$s = $fc->pbacktrack(100, RNA::PBACKTRACK_NON_REDUNDANT);
ok(scalar(@{$s}) == 100);

$failure = 0;
foreach my $s (@{$s}) {
  if (length($s) != length($sequence)) {
    $failure = 1;
    last;
  }
}
ok($failure == 0);

# check for uniqueness, i.e. no duplicates
@ss = uniq(@{$s});
ok(scalar(@{$s}) == scalar(@ss));


print "test_pbacktrack_nr_resume\n";
$d = undef;
@ss = ();

foreach my $i (1..$iterations) {
  ($d, $s) = $fc->pbacktrack($num_samples, $d, RNA::PBACKTRACK_NON_REDUNDANT);
  push(@ss, @{$s});
}

ok(scalar(@ss) == ($iterations * $num_samples));

$failure = 0;
foreach my $s (@ss) {
  if (length($s) != length($sequence)) {
    $failure = 1;
    last;
  }
}
ok($failure == 0);

# check for uniqueness, i.e. no duplicates
@sss = uniq(@ss);
ok(scalar(@ss) == scalar(@sss));


print "test_pbacktrack5_cb (multiple sub-structures, a.k.a. pbacktrack5_cb)\n";
@ss = ();
$i  = $fc->pbacktrack5(20, 10, \&store_structure, \@ss);
ok($i == 20);
ok(scalar(@ss) == 20);

$failure = 0;
foreach my $s (@ss) {
  if (length($s) != 10) {
    $failure = 1;
    last;
  }
}
ok($failure == 0);

@ss = ();
$i  = $fc->pbacktrack5(100, 50, \&store_structure, \@ss);
ok($i == 100);
ok(scalar(@ss) == 100);

$failure = 0;
foreach my $s (@ss) {
  if (length($s) != 50) {
    $failure = 1;
    last;
  }
}
ok($failure == 0);


print "test_pbacktrack_cb  (multiple structures, a.k.a. pbacktrack_cb)\n";
@ss = ();
$i  = $fc->pbacktrack(100, \&store_structure, \@ss);
ok($i == 100);
ok(scalar(@ss) == 100);

$failure = 0;
foreach my $s (@ss) {
  if (length($s) != length($sequence)) {
    $failure = 1;
    last;
  }
}
ok($failure == 0);


print "test_pbacktrack_nr_cb\n";
@ss = ();
$i  = $fc->pbacktrack(100, \&store_structure, \@ss, RNA::PBACKTRACK_NON_REDUNDANT);
ok($i == 100);
ok(scalar(@ss) == 100);

$failure = 0;
foreach my $s (@ss) {
  if (length($s) != length($sequence)) {
    $failure = 1;
    last;
  }
}
ok($failure == 0);

# check for uniqueness, i.e. no duplicates
@sss = uniq(@ss);
ok(scalar(@ss) == scalar(@sss));


print "test_pbacktrack_nr_resume_cb\n";
$d  = undef;
@ss = ();

$failure = 0;
foreach my $i (1..$iterations) {
  ($d, $i) = $fc->pbacktrack($num_samples, \&store_structure, \@ss, $d, RNA::PBACKTRACK_NON_REDUNDANT);
  $failure = 1, last if $i != $num_samples;
}
ok($failure == 0);

ok(scalar(@ss) == ($iterations * $num_samples));

$failure = 0;
foreach my $s (@ss) {
  if (length($s) != length($sequence)) {
    $failure = 1;
    last;
  }
}
ok($failure == 0);

# check for uniqueness, i.e. no duplicates
@sss = uniq(@ss);
ok(scalar(@ss) == scalar(@sss));
